sql: disallow creating an index with a REFCURSOR column

internal/sqlsmith.TestRandTableInserts [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12126752?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12126752?buildTab=artifacts#/) on master @ [45882fb8aef5ac61b13de9bd2d5ea59a21f88510](https://github.com/cockroachdb/cockroach/commits/45882fb8aef5ac61b13de9bd2d5ea59a21f88510):


```
=== RUN   TestRandTableInserts
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/f82616d5c9a8c3209014f997587365ae/logTestRandTableInserts4216486116
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:124: automatically injected virtual cluster under test; see comment at top of test_server_shim.go for details.
    sqlsmith_test.go:58: ALTER TABLE table3 ADD FOREIGN KEY ("col3*_4") REFERENCES table10 (col10_2) ON DELETE CASCADE ON UPDATE CASCADE
    sqlsmith_test.go:59: pq: validate fk constraint: unsupported comparison operator: <refcursor> = <refcursor>
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f82616d5c9a8c3209014f997587365ae/logTestRandTableInserts4216486116
--- FAIL: TestRandTableInserts (4.17s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandTableInserts.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-32234

--- a/pkg/sql/catalog/colinfo/col_type_info.go
+++ b/pkg/sql/catalog/colinfo/col_type_info.go
@@ -150,7 +150,7 @@ func ValidateColumnDefType(ctx context.Context, version clusterversion.Handle, t
 
 // ColumnTypeIsIndexable returns whether the type t is valid as an indexed column.
 func ColumnTypeIsIndexable(t *types.T) bool {
-	if t.IsAmbiguous() || t.Family() == types.TupleFamily {
+	if t.IsAmbiguous() || t.Family() == types.TupleFamily || t.Family() == types.RefCursorFamily {
 		return false
 	}
 	// Some inverted index types also have a key encoding, but we don't

--- a/pkg/sql/logictest/testdata/logic_test/refcursor
+++ b/pkg/sql/logictest/testdata/logic_test/refcursor
@@ -553,4 +553,54 @@ SELECT 'foo'::REFCURSOR IS DISTINCT FROM (CASE WHEN true THEN NULL ELSE NULL::RE
 statement error pgcode 22023 pq: unsupported comparison operator
 SELECT 'foo'::REFCURSOR IS NOT DISTINCT FROM (CASE WHEN true THEN NULL ELSE NULL::REFCURSOR END);
 
+# Regression test for #112099 - REFCURSOR is not valid as an index column.
+subtest refcursor_index
+
+statement ok
+CREATE TABLE t112099_no_index (x INT NOT NULL, y TEXT NOT NULL, r REFCURSOR NOT NULL);
+
+# REFCURSOR is not allowed in a primary key.
+statement error pgcode 0A000 pq: unimplemented: column r is of type refcursor and thus is not indexable
+CREATE TABLE t112099 (r REFCURSOR PRIMARY KEY);
+
+statement error pgcode 0A000 pq: unimplemented: column r is of type refcursor and thus is not indexable
+CREATE TABLE t112099 (x INT, y TEXT, r REFCURSOR, PRIMARY KEY (x, r, y));
+
+statement error pgcode 0A000 pq: unimplemented: column r is of type refcursor and thus is not indexable
+ALTER TABLE t112099_no_index ADD PRIMARY KEY (r);
+
+statement error pgcode 0A000 pq: unimplemented: column r is of type refcursor and thus is not indexable
+ALTER TABLE t112099_no_index ADD PRIMARY KEY (x, r, y);
+
+# REFCURSOR is not allowed in a secondary index.
+statement error pgcode 0A000 pq: unimplemented: column r is of type refcursor and thus is not indexable
+CREATE TABLE t112099 (r REFCURSOR, INDEX (r));
+
+statement error pgcode 0A000 pq: unimplemented: column r is of type refcursor and thus is not indexable
+CREATE TABLE t112099 (x INT, y TEXT, r REFCURSOR, INDEX (x, r, y));
+
+statement error pgcode 0A000 pq: unimplemented: column r is of type refcursor and thus is not indexable
+CREATE INDEX ON t112099_no_index (r);
+
+statement error pgcode 0A000 pq: unimplemented: column r is of type refcursor and thus is not indexable
+CREATE INDEX ON t112099_no_index (x, r, y);
+
+# REFCURSOR is allowed as a STORING column.
+statement ok
+CREATE TABLE t112099 (x INT, r REFCURSOR, INDEX (x) STORING (r));
+DROP TABLE t112099;
+
+statement ok
+CREATE TABLE t112099 (x INT, y TEXT, r REFCURSOR, INDEX (x) STORING (y, r));
+DROP TABLE t112099;
+
+statement ok
+CREATE INDEX ON t112099_no_index (x) STORING (r);
+
+statement ok
+CREATE INDEX ON t112099_no_index (x) STORING (y, r);
+
+statement ok
+DROP TABLE t112099_no_index;
+
 subtest end

